src/backend/gporca/libgpopt/src/base/CUtils.cpp

```diff
@@ -903,21 +903,27 @@ CUtils::FHasCTEAnchor(CExpression *pexpr)
 // return CTEConsumers' and a set of CTEProducers' CTE ids in the given subtree
 void
 CUtils::CollectConsumersAndProducers(CMemoryPool *mp, CExpression *pexpr,
-									 ULongPtrArray *cteConsumers,
-									 UlongCteIdHashSet *cteProducerSet)
+									 UlongCteIdHashSet *cteConsumers,
+									 UlongCteIdHashSet *cteProducerSet,
+									 BOOL isRoot)
 {
 	COperator *pop = pexpr->Pop();
 
 	if (COperator::EopPhysicalCTEConsumer == pexpr->Pop()->Eopid())
 	{
-		cteConsumers->Append(GPOS_NEW(mp) ULONG(
+		cteConsumers->Insert(GPOS_NEW(mp) ULONG(
 			CPhysicalCTEConsumer::PopConvert(pop)->UlCTEId()));
 	}
 	else if (COperator::EopPhysicalCTEProducer == pexpr->Pop()->Eopid())
 	{
 		cteProducerSet->Insert(GPOS_NEW(mp) ULONG(
 			CPhysicalCTEProducer::PopConvert(pop)->UlCTEId()));
 	}
+	else if (FIsAMotionOperator(pexpr->Pop()->Eopid()) && !isRoot)
```

Comment:
Why not inline the condition, as it is done, for example in https://github.com/arenadata/gpdb/blob/4fa8a09d5babb471a36b8a06b499247d24d1f6ad/src/backend/gporca/libgpdbcost/src/CCostModelGPDB.cpp#L1449-L1454

Reply:
Approach with usage of function, seems more unreadable for me, also this function may be reused after

Reply:
I think, such function already implemented, see `CUtils::FPhysicalMotion`. Maybe reuse it?

Reply:
Yes, I've searched functions that checks OperatorId,  and forgot that we can check motion with other approach